platform/linux-generic/odp_crypto.c

```diff
@@ -308,6 +318,182 @@ void packet_hmac(odp_packet_t pkt,
 	HMAC_Final(ctx, hash, NULL);
 }
 
+static void do_pad_xor(uint8_t *out, const uint8_t *in, int len)
+{
+	int pos = 0;
+
+	for (pos = 0; pos < AES_BLOCK_SIZE; pos++, in++, out++) {
+		if (pos < len)
+			*out ^= *in;
+		if (pos >= len) {
+			*out ^= 0x80;
+			break;
+		}
+	}
+}
+
+static void xor_block(uint8_t *res, const uint8_t *op)
+{
+	int i;
+
+	for (i = 0; i < AES_BLOCK_SIZE; i++)
+		res[i] ^= op[i];
+}
+
+static
+void packet_aes_xcbc_mac(odp_packet_t pkt,
+			 const odp_crypto_packet_op_param_t *param,
+			 odp_crypto_generic_session_t *session,
+			 uint8_t *hash)
+{
+	uint8_t e[AES_BLOCK_SIZE] = {0};
+	uint8_t seg[AES_BLOCK_SIZE] = {0};
+	bool is_seg = false;
+	uint32_t seg_datalen = 0;
+	uint32_t offset = param->auth_range.offset;
+	uint32_t len   = param->auth_range.length;
+	uint32_t seglen = 0;
+	uint32_t datalen = 0;
+	int dummy_len = 0;
+	EVP_CIPHER_CTX *ctx;
+	void *mapaddr;
+	uint8_t *data = NULL;
+
+	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
+	ODP_ASSERT(session != NULL);
+	ODP_ASSERT(sizeof(session->auth.key) >= 3 * AES_KEY_LENGTH);
+
+	ctx = EVP_CIPHER_CTX_new();
+	EVP_EncryptInit_ex(ctx, session->auth.evp_cipher,
+			   NULL, session->auth.key, NULL);
+
+	while (len > 0) {
+		mapaddr = odp_packet_offset(pkt, offset, &seglen, NULL);
+		datalen = seglen >= len ? len : seglen;
+		data = (uint8_t *)mapaddr;
+		offset += datalen;
+		len -= datalen;
+		/* Segmentation handle */
+		if (is_seg) {
+			is_seg = false;
+			if (datalen > AES_BLOCK_SIZE - seg_datalen) {
+				memcpy(seg + seg_datalen,
+				       data, AES_BLOCK_SIZE - seg_datalen);
+				xor_block(e, seg);
+				EVP_EncryptUpdate(ctx,
+						  e, &dummy_len, e, sizeof(e));
+				datalen -= AES_BLOCK_SIZE - seg_datalen;
+			} else {
+				memcpy(seg + seg_datalen, data, datalen);
+				data = seg;
+				datalen += seg_datalen;
```

Comment:
This will result in `memcpy(seg, seg, datalen)` later. I suggest to have the following code in this clause:

```c
				memcpy(seg + seg_datalen, data, datalen);
				seg_datalen += datalen;
				continue;
```

Reply:
@lumag if the code run into the else logic, it means that's the last segment and it will not enter the memcpy.  it will run do_pad_xor(e, data, datalen); So the datalen must be the last block length.
Your change looks good but the datalen is not correct, I think.
If you are free, I'd like to talk to you for this part now.

Reply:
If there's one test case for the segment packet, it will give great help for this part.

Reply:
@SonicwallYhe it might be a segment, that is just less than AES_BLOCK_SIZE (quite theoretical thing, but it might become real because of packet references).
I'm considering the following scenario:
 - you have previous partial segment
 - new segment is not enough to fill data up to full AES block

In this case you don't have to go through for loop and the following if. You can just shortcut it here.

Reply:
@lumag Yes, if this is a segment, it means that's the last segment.(it's impossible for a packet whose's length is less than 16 bytes). If it's last segment, the var len == 0. 
In this case, we can just break the loop.

Reply:
@SonicwallYhe it might be an interim short segment .

Reply:
@lumag Then what's your advice about the code here?

Reply:
@SonicwallYhe 
```c
static void memxor(uint8_t *res, const uint8_t *op, size_t len)
{
	for (size_t i = 0; i < len; i++)
		res[i] ^= op[i];
}

static
void packet_aes_xcbc_mac(odp_packet_t pkt,
		 const odp_crypto_packet_op_param_t *param,
		 odp_crypto_generic_session_t *session,
		 uint8_t *hash)
{
	uint8_t e[AES_BLOCK_SIZE] = {0};
	size_t eoff = 0;
	uint32_t offset = param->auth_range.offset;
	uint32_t len   = param->auth_range.length;
	uint32_t seglen = 0;
	uint32_t datalen = 0;
	int dummy_len = 0;
	EVP_CIPHER_CTX *ctx;
	void *mapaddr;
	uint8_t *data = NULL;

	ODP_ASSERT(offset + len <= odp_packet_len(pkt));
	ODP_ASSERT(session != NULL);
	ODP_ASSERT(sizeof(session->auth.key) >= 3 * AES_KEY_LENGTH);

	ctx = EVP_CIPHER_CTX_new();
	EVP_EncryptInit_ex(ctx, session->auth.evp_cipher,
			   NULL, session->auth.key, NULL);

	while (len > 0) {
		mapaddr = odp_packet_offset(pkt, offset, &seglen, NULL);
		datalen = seglen >= len ? len : seglen;
		data = (uint8_t *)mapaddr;
		offset += datalen;
		len -= datalen;
		if (eoff != 0) {
			if (eoff + datalen > AES_BLOCK_SIZE) {
				memxor(e + eoff, data, AES_BLOCK_SIZE - eoff);
				datalen -= (AES_BLOCK_SIZE - eoff);
				eoff = 0;
				EVP_EncryptUpdate(ctx,
						  e, &dummy_len, e, sizeof(e));
			} else {
				memxor(e + eoff, data, datalen);
				eoff += datalen;
				continue;
			}
		}
		while (datalen > AES_BLOCK_SIZE) {
			xor_block(e, data);
			EVP_EncryptUpdate(ctx, e, &dummy_len, e, sizeof(e));
			data += AES_BLOCK_SIZE;
			datalen -= AES_BLOCK_SIZE;
		}
		/* Segmentation handle */
		if (datalen > 0) {
			memxor(e, data, datalen);
			eoff = datalen;
		}
	}

	if (eoff == AES_BLOCK_SIZE) {
		xor_block(e, session->auth.key + AES_KEY_LENGTH);
	} else {
		e[eoff] ^= 0x80;
		xor_block(e, session->auth.key + AES_KEY_LENGTH * 2);
	}
	EVP_EncryptUpdate(ctx, hash, &dummy_len, e, sizeof(e));
	EVP_CIPHER_CTX_free(ctx);
}
```